File: common/src/main/java/juuxel/adorn/util/animation/AbstractAnimatedProperty.java
package juuxel.adorn.util.animation;

import org.jspecify.annotations.Nullable;

import java.util.Objects;

public abstract class AbstractAnimatedProperty<T> {
    private final AnimationEngine engine;
    private final int duration;
    private final Interpolator<T> interpolator;
    private @Nullable Task currentTask = null;

    protected AbstractAnimatedProperty(AnimationEngine engine, int duration, Interpolator<T> interpolator) {
        this.engine = engine;
        this.duration = duration;
        this.interpolator = interpolator;
    }

    protected abstract void setRawValue(T value);

    public abstract T get();

    public synchronized void set(T value) {
        T oldValue = get();
        var oldTask = currentTask;
        if (oldTask != null) engine.remove(oldTask);

        if (!Objects.equals(oldValue, value)) {
            var task = new Task(oldValue, value);
            currentTask = task;
            engine.add(task);
        }
    }

    // https://easings.net/#easeOutQuint
    private static float ease(float delta) {
        return 1 - (float) Math.pow(1 - delta, 5);
    }

    private final class Task implements AnimationTask {
        private final T from;
        private final T to;
        private int age = 0;

        private Task(T from, T to) {
            this.from = from;
            this.to = to;
        }

        @Override
        public boolean isAlive() {
            return age < duration;
        }

        @Override
        public void tick() {
            age++;
            float delta = ease((float) age / (float) duration);
            T newValue = interpolator.interpolate(delta, from, to);
            setRawValue(newValue);
        }

        @Override
        public void removed() {
            synchronized (AbstractAnimatedProperty.this) {
                if (currentTask == this) {
                    currentTask = null;
                }
            }
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/util/animation/AbstractAnimatedProperty.kt
package juuxel.adorn.util.animation

import kotlin.math.pow
import kotlin.properties.ReadWriteProperty
import kotlin.reflect.KProperty

abstract class AbstractAnimatedProperty<T>(
    private val engine: AnimationEngine,
    private val duration: Int,
    private val interpolator: Interpolator<T>
) : ReadWriteProperty<Any?, T> {
    protected abstract fun setRawValue(value: T)

    @Volatile
    private var currentTask: Task? = null

    override fun setValue(thisRef: Any?, property: KProperty<*>, value: T) {
        synchronized(this) {
            val oldValue = getValue(thisRef, property)
            currentTask?.let(engine::remove)

            if (oldValue != value) {
                val task = Task(oldValue, value)
                currentTask = task
                engine.add(task)
            }
        }
    }

    companion object {
        // https://easings.net/#easeOutQuint
        private fun ease(delta: Float): Float =
            1 - (1 - delta).pow(5)
    }

    private inner class Task(private val from: T, private val to: T) : AnimationTask {
        private var age = 0

        override fun isAlive(): Boolean = age < duration

        override fun tick() {
            age++
            val delta = ease(age.toFloat() / duration.toFloat())
            val newValue = interpolator.interpolate(delta, from, to)
            setRawValue(newValue)
        }

        override fun removed() {
            synchronized(this@AbstractAnimatedProperty) {
                if (currentTask == this) {
                    currentTask = null
                }
            }
        }
    }
}
